src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java

```diff
@@ -2722,6 +2758,96 @@ public void validateDeleteRequest(
   }
 
+  public List<String> exportTableToParquetWithGsPaths(
+      Snapshot snapshot, GoogleBucketResource bucketResource, String flightId)
+      throws InterruptedException {
+    List<String> paths = new ArrayList<>();
+    BigQueryProject bigQueryProject = BigQueryProject.from(snapshot);
+
+    for (var table : snapshot.getTables()) {
+      String tableName = table.getName();
+      String exportPath =
+          new ST(exportPathTemplate)
+              .add("bucket", bucketResource.getName())
+              .add("flightId", flightId)
+              .add("table", tableName)
+              .render();
+
+      String exportToParquetQuery = createExportToParquetWithGsPathQuery(snapshot, table, flightId);
+
+      String exportStatement =
+          new ST(exportToParquetTemplate)
+              .add("exportPath", exportPath)
+              .add("exportToParquetQuery", exportToParquetQuery)
+              .add("table", tableName)
+              .render();
+
+      bigQueryProject.query(exportStatement);
+      paths.add(exportPath);
+    }
+
+    return paths;
+  }
+
+  private static final String exportToMappingTableTemplate =
+      "WITH datarepo_gs_path_mapping AS "
+          + "(SELECT <fileIdKey>, <gsPathKey> "
+          + "  FROM `<project>.<snapshotDatasetName>.<gsPathMappingTable>`) "
+          + "SELECT "
+          + "  S.<pdaoRowIdColumn>,"
+          + "  <mappedColumns; separator=\",\"> "
+          + "FROM "
+          + "`<project>.<snapshotDatasetName>.<table>` S";
+
+  private String createExportToParquetWithGsPathQuery(
+      Snapshot snapshot, SnapshotTable table, String flightId) {
+    String gsPathMappingTableName = BigQueryUtils.gsPathMappingTableName(snapshot);
+    String extTableName = externalTableName(gsPathMappingTableName, flightId);
+    List<String> mappedColumns =
+        table.getColumns().stream().map(this::gsPathMappingSelectSql).collect(Collectors.toList());
+
+    return new ST(exportToMappingTableTemplate)
+        .add("project", snapshot.getProjectResource().getGoogleProjectId())
+        .add("snapshotDatasetName", snapshot.getName())
+        .add("gsPathMappingTable", extTableName)
+        .add("pdaoRowIdColumn", PDAO_ROW_ID_COLUMN)
+        .add("mappedColumns", mappedColumns)
+        .add("table", table.getName())
+        .add("gsPathKey", PDAO_FIRESTORE_DUMP_GSPATH_KEY)
+        .add("fileIdKey", PDAO_FIRESTORE_DUMP_FILE_ID_KEY)
+        .render();
+  }
+
+  private static final String fileRefMappingColumnTemplate =
+      "(SELECT <gsPathKey> "
+          + "FROM datarepo_gs_path_mapping "
+          + "WHERE RIGHT(S.<columnName>, 36) = <fileIdKey>) AS <columnName>";
+
+  private static final String fileRefArrayOfMappingColumnTemplate =
+      "  ARRAY(SELECT <gsPathKey> "
+          + "    FROM UNNEST(<columnName>) AS unnested_drs_path, "
+          + "    datarepo_gs_path_mapping "
+          + "    WHERE RIGHT(unnested_drs_path, 36) = <fileIdKey>) AS <columnName>";
```

Comment:
It's a little scary that there's no way to validate that the DRS ID is in the format that we expect but I'm not sure that there's much else we can do

Reply:
Yeah, I don't love it, but I'm ok with it, since there's no chance of data loss here. Export just breaks if the format changes, but the data is still there and exportable in DRS format. This also basically undoes the snapshot_id+file_id concatenation on snapshot creation, so it makes sense. I experimented with other ways, but this is the most concise and easiest to understand.